### PR TITLE
EASY-1496 File content indexing of images fails because of timestamp formats

### PR DESCRIPTION
**Pull request by janvanmansum**
Fixes EASY-1496

#### When applied it will
* Make the mime types for which to attempt content extraction configurable.

#### Related pull requests on github

To do:
- [x] Fix unit tests
- [x] Test in `easy-dtap`
- [x] Refactor setting names
- [x] Document settings

### CODE REVIEW

**Review by rvanheest — APPROVED** [review]

File: src/main/assembly/dist/cfg/application.properties [hunk 1]
```diff
@@ -11,5 +11,13 @@ ldap.provider.url=ldap://localhost
...
-# 64*1024*1024
-max-fileSize-toExtract-content-from=67108864
\ No newline at end of file
+#
+# The service will try to have Solr index extracted content from the file. This is limited to the
+# media types and file size configured below.
+#
+file-content-extraction.max-filesize=67108864
+file-content-extraction.mime-types=\
+  text/plain, \
```
Comment (janvanmansum) [inline src/main/assembly/dist/cfg/application.properties:20]: @PaulBoon , @jo-pol : do we need more default media types here?

Comment (jo-pol) [reply src/main/assembly/dist/cfg/application.properties:20]: In this list I don't see pdf, in the `application.properties` I do see pdf but otherwise much less.

File: src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala [hunk 2]
```diff
@@ -49,5 +49,8 @@ trait ApplicationWiring
   override val solrClient: SolrClient = new HttpSolrClient.Builder(solrUrl.toString).build()
   override val vaultBaseUri: URI = new URI(configuration.properties.getString("vault.url", "http://localhost"))
   //TODO BagStoreComponent using HttpWorker as in easy-download
-  override val maxFileSizeToExtractContentFrom: Double = configuration.properties.getString("max-fileSize-toExtract-content-from", (64 * 1024 * 1024).toString).toDouble
+
+  // Cannot use properties.getList in the following line because we do not parse commas. Parsing commas conflicts with having LDAP DNs as values.
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:53]: You are lucky that mime types don't have commas as far as I know. Another option would be to parse comma's and join the LDAP DN value components.

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala [hunk 3]
```diff
@@ -55,19 +56,21 @@ trait Solr extends DebugEnhancedLogging {
...
-    else {
+    val fileSizeOkForIndexing = maxFileSizeToExtractContentFrom > item.size
+    val fileMimeTypeEligibleForIndexing = mimeTypesToExtractContentFrom.contains(item.mimeType)
+    if (fileSizeOkForIndexing && fileMimeTypeEligibleForIndexing) {
       logger.info(s"Submission with content of [$solrDocId]")
       submitWithContent(fileUrl, solrDocId, solrLiterals)
         .map(_ => FileSubmittedWithContent(solrDocId))
         .recoverWith { case t =>
-          logger.warn(s"Submission with content of [$solrDocId] failed with ${ t.getMessage }")
+          logger.warn(s"Submission with content of [$solrDocId] failed. Proceeding to index only metadata. Message for Solr: ${ t.getMessage }")
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala]: for Solr -> from Solr

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala [hunk 4]
```diff
@@ -55,19 +56,21 @@ trait Solr extends DebugEnhancedLogging {
...
       )
     }
-    if (item.size > maxFileSizeToExtractContentFrom) {
-      logger.info(s"Submission without content of [$solrDocId]")
-      submitWithOnlyMetadata(solrDocId, solrLiterals).map(_ => FileSubmittedWithOnlyMetadata(solrDocId))
-    }
-    else {
+    val fileSizeOkForIndexing = maxFileSizeToExtractContentFrom > item.size
+    val fileMimeTypeEligibleForIndexing = mimeTypesToExtractContentFrom.contains(item.mimeType)
+    if (fileSizeOkForIndexing && fileMimeTypeEligibleForIndexing) {
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:61]: Perhaps extract method for this large branch if it doesn't cause too many arguments.

File: src/test/scala/nl.knaw.dans.easy.solr4files/AppSpec.scala [hunk 5]
```diff
@@ -90,9 +90,9 @@ class AppSpec extends TestSupportFixture {
     val app = new StubbedSolrApp()
     Seq(
       "data/path/to/a/random/video/hubble.mpg",
+      "data/reisverslag/centaur-nederlands.srt",
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.solr4files/AppSpec.scala:93]: Just curious: why did you need this change?

File: src/test/scala/nl.knaw.dans.easy.solr4files/AppSpec.scala [hunk 6]
```diff
@@ -127,7 +127,7 @@ class AppSpec extends TestSupportFixture {
 
     val result = app.update(storeName, uuidAnonymized)
     inside(result) { case Success(feedback) =>
-      feedback.toString shouldBe s"Bag ${ uuidAnonymized }: 3 times FileSubmittedWithContent"
+      feedback.toString shouldBe s"Bag ${ uuidAnonymized }: 2 times FileSubmittedWithContent, 1 times FileSubmittedWithOnlyMetadata"
```
Comment (jo-pol) [inline src/test/scala/nl.knaw.dans.easy.solr4files/AppSpec.scala:130]: 👍 to show both numbers

**Review by jo-pol — COMMENTED** [review]
Use the solr query interface to see what mime-types we have so far for inspiration. Not sure how many subtle variations we might encounter.

A complete list in the hard coded default may be less important. Note that updates made in the configuration of the production system sooner or later might get lost when releasing, so please keep the configuration in the code up to date.

File: src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala (hunk 2 above)
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:53]: Yes, I did think to check that first ;)

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala (hunk 3 above)
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala]: 👍

File: src/main/assembly/dist/cfg/application.properties (hunk 1 above)
Comment (janvanmansum) [reply src/main/assembly/dist/cfg/application.properties:20]: Confusing comment. This file *is* `application.properties`. You are probably referring to `debug-config/application.properties` specifically. I'll add pdf and make this properties the same in both files.

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala (hunk 4 above)
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:61]: Decided against it, as the the log lines document pretty succinctly what we are doing here.

File: src/test/scala/nl.knaw.dans.easy.solr4files/AppSpec.scala (hunk 5 above)
Comment (janvanmansum) [reply src/test/scala/nl.knaw.dans.easy.solr4files/AppSpec.scala:93]: Not sure if it is needed. I just made the order the same as in the corresponding `files.xml`.

**Comment by PaulBoon** [comment]
Hmm, already merged @janvanmansum @jo-pol . 
Maybe we should have done a 'blacklist' approach instead of a 'whitelist' approach?
Definitely we want pdf, csv, and xml in the whitelist.
